sql/src/main/java/io/crate/operation/fetch/NodeFetchOperation.java

```diff
@@ -21,190 +21,91 @@
+            // if ref is null, there is nothing to fetch for this relation
+            if (ref != null) {
+                TableFetchInfo tableFetchInfo = new TableFetchInfo(references, fetchContext);
+                result.put(ref.ident().tableIdent(), tableFetchInfo);
             }
-
-            @Override
-            public void onFailure(@Nonnull Throwable t) {
-                fetchContext.close(); // closeDueToFailure(t);
-            }
-        });
-
-        List<LuceneDocFetcher> shardFetchers = new ArrayList<>(numShards);
-        for (IntObjectCursor<ShardDocIdsBucket> entry : shardBuckets) {
-            Engine.Searcher searcher = fetchContext.searcher(entry.key);
-            IndexService indexService = fetchContext.indexService(entry.key);
-            // create new collect expression for every shard (collect expressions are not thread-safe)
-            CollectInputSymbolVisitor.Context docCtx = docInputSymbolVisitor.extractImplementations(toFetchReferences);
-            shardFetchers.add(
-                    new LuceneDocFetcher(
-                            docCtx.topLevelInputs(),
-                            docCtx.docLevelExpressions(),
-                            upstreamsRowMerger,
-                            entry.value,
-                            indexService.mapperService(),
-                            indexService.fieldData(),
-                            searcher,
-                            fetchContext));
         }
-        try {
-            runFetchThreaded(shardFetchers, ramAccountingContext);
-        } catch (RejectedExecutionException e) {
-            bucketBuilder.fail(e);
-        }
-
-        if (LOGGER.isTraceEnabled()) {
-            LOGGER.trace("started {} shardFetchers", numShards);
-        }
-    }
-
-    private void runFetchThreaded(final List<LuceneDocFetcher> shardFetchers,
-                                  final RamAccountingContext ramAccountingContext) throws RejectedExecutionException {
-
-        ThreadPools.runWithAvailableThreads(
-                executor,
-                poolSize,
-                Lists.transform(shardFetchers, new Function<LuceneDocFetcher, Runnable>() {
-
-                    @Nullable
-                    public Runnable apply(final LuceneDocFetcher input) {
-                        return new Runnable() {
-                            @Override
-                            public void run() {
-                                input.doFetch(ramAccountingContext);
-                            }
-                        };
-                    }
-                })
-        );
+        // this class should not be used if no fetches will be requested
+        assert !result.isEmpty();
+        return result;
     }
 
-    static class ShardDocIdsBucket {
-
-        private final IntArrayList positions = new IntArrayList();
-        private final IntArrayList docIds = new IntArrayList();
+    public IntObjectOpenHashMap<StreamBucket> doFetch(
```

Comment:
would it be possible to only return IntObjectMap here? to avoid relying on implementation detail